[HttpClientFactory] Do not log query string by default

antonfirsov opened the pull request:
Since query strings often contain sensitive information, we decided to avoid including them in HttpClientFactory logs by default. For use-cases where logging query string is necessary and safe, query string logging can be turned on globally by an AppContext switch `System.Net.Http.DisableUriRedaction` .

In .NET 10 we plan to support more sophisticated redaction scenarios.

antonfirsov committed:
Do not log query strings in HttpClientFactory

antonfirsov committed:
shuffle tests

antonfirsov commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -88,6 +89,8 @@ protected override HttpResponseMessage Send(HttpRequestMessage request, Cancella
         // Used in tests.
         internal static class Log
         {
+            private static readonly bool s_logQueryString = AppContext.TryGetSwitch("Microsoft.Extensions.Http.LogQueryString", out bool value) && value;
```
The other name option I considered is `EnableQueryStringLogging`. Which one is better?

liveans replied:
It feels like `EnableQueryStringLogging` is better.

CarnaViire replied:
I'd actually vote for using the same terminology that's _already used_ in dotnet/extensions logging (i.e. "_redaction_"), and also because it's a "_rollback_" switch instead of an "experimental feature" switch, I'd vote for saying "_disable_" instead of "enable":

`Microsoft.Extensions.Http.DisableUriRedaction`
or
`Microsoft.Extensions.Http.DisableUriQueryRedaction`

P.S.: We have at least 2 examples for similar "rollback" switches, and they also use "disable" wording: `System.Net.Security.DisableTlsResume` and `System.Net.Quic.DisableConfigurationCache`

P.P.S.: I also believe we've always added an _environmental variable_ alongside the AppContext switch (e.g. `DOTNET_SYSTEM_NET_SECURITY_DISABLETLSRESUME`) -- I think we should add one here as well.

antonfirsov replied:
> We have at least 2 examples for similar "rollback" switches, and they also use "disable" wording: `System.Net.Security.DisableTlsResume` and `System.Net.Quic.DisableConfigurationCache`

From what I see, both switches have opposite semantics: a feature is enabled by default, and you need to explicitly assign a `true` to the switch value in-order to disable it. Same is true for `System.Net.SocketsHttpHandler.Http2FlowControl.DisableDynamicWindowSizing`.

In this case, query string logging would be disabled by default, so one would need to assign  `DisableUriQueryRedaction = false` in order to enable it. It might be just me, but it feels more confusing  than having an *Enable* switch that has to be assigned `true` in order to opt-in into the feature.

Examples for the logic I prefer: `System.ComponentModel.TypeConverter.EnableUnsafeBinaryFormatterInDesigntimeLicenseContextSerialization`,  `System.Diagnostics.DefaultActivityIdFormatIsHierarchial`, `DOTNET_SYSTEM_NET_SOCKETS_INLINE_COMPLETIONS`, `DOTNET_SYSTEM_RUNTIME_CACHING_TRACING`, `System.Resources.UseSystemResourceKeys` and many more.

CarnaViire replied:
> From what I see, both switches have opposite semantics: a feature is enabled by default, and you need to explicitly assign a true to the switch value in-order to disable it.

But this is exactly what we have here, isn't it? Redaction = removing or substituting query string with "*". It _is_ enabled by default (with the change). If I want to "rollback" - to _stop redacting_ and see raw unredacted uris in my logs -- I say `DisableUriQueryRedaction = true`

antonfirsov replied:
> `DisableUriQueryRedaction = true`

Ah sorry, that makes sense! Got lost a bit with the negations 😅

MihaZupan commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -144,11 +147,15 @@ public static void RequestEnd(ILogger logger, HttpResponseMessage response, Time
+                    return null;
+                }
+
+                Debug.Assert(uri.IsAbsoluteUri, "HttpClient should not forward relative Uri to its handlers.");
+                return s_logQueryString ? uri.AbsoluteUri : $"{uri.Scheme}://{uri.Host}{uri.AbsolutePath}";
```
```suggestion
                if (uri.IsAbsoluteUri)
                {
                    return s_logQueryString
                        ? uri.AbsoluteUri
                        : uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.Query, UriFormat.UriEscaped);
                }

                string uriString = uri.ToString();
                int queryOffset = s_logQueryString ? -1 : uriString.IndexOf('?');
                return queryOffset < 0
                    ? uriString
                    : uriString.Substring(0, queryOffset);
```

The handler is public, so it's possible for this to still be a relative Uri https://github.com/dotnet/runtime/pull/63012#discussion_r781199208.

`GetComponents` can save you the intermediate string allocation for the path.

antonfirsov committed:
Update GetUriString

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>

antonfirsov committed:
add tests for relative Uri logging

antonfirsov committed:
remove unnecessary using

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingScopeHttpMessageHandler.cs:
```diff
@@ -113,12 +113,12 @@ public static class EventIds
             public static IDisposable? BeginRequestPipelineScope(ILogger logger, HttpRequestMessage request)
             {
-                return _beginRequestPipelineScope(logger, request.Method, GetUriString(request.RequestUri));
+                return _beginRequestPipelineScope(logger, request.Method, LoggingHttpMessageHandler.Log.GetUriString(request.RequestUri));
```
Let's not use one handler from inside of the unrelated other? Given the (size of the) change @MihaZupan suggested, I think we might need a separate `LogHelper`/`RedactionHelper` class or something

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LoggingHttpMessageHandler.cs:
```diff
@@ -144,11 +147,25 @@ public static void RequestEnd(ILogger logger, HttpResponseMessage response, Time
+                int queryOffset = s_logQueryString ? -1 : uriString.IndexOf('?');
+                return queryOffset < 0
+                    ? uriString
+                    : uriString.Substring(0, queryOffset);
```
I've been thinking that we might want to leave a "hint" that we've redacted something here. dotnet/extensions logging uses `REDACTED`; but our own header redaction in HttpClientFactory substitutes header values with `*` -- so I've been thinking that we need to align and also substitute the query with `*`?

Also given the headers are not removed, but substituted instead, it also seems more suitable to substitute (instead of removing)

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/LoggingUriOutputTests.cs:
```diff
@@ -16,156 +16,151 @@ namespace Microsoft.Extensions.Http.Tests.Logging
-        [Fact]
-        public async Task LoggingScopeHttpMessageHandler_LogsAbsoluteUri()
+        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        [MemberData(nameof(Handlers_LogAbsoluteUri_Data))]
```
```suggestion
        [MemberData(nameof(Handlers_LogExpectedUri_Data))]
```
to align with test name

CarnaViire reviewed (changes requested):
See comments above

stephentoub commented:
We don't want to expose a programmatic option for this on HttpClientFactoryOptions or somewhere similar?

antonfirsov commented:
> We don't want to expose a programmatic option for this on HttpClientFactoryOptions or somewhere similar?

We weren't able to agree on an API for this an related requests, like support for [`url.template`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#experimental-attributes), Activity enrichment and filtering.  

Many customers (including our first-parties) would find the usability of a HttpClientFactory-scoped Uri-redaction API limited without these features. In order to deliver a consistent solution for all the requests in the telemetry space, we decided to push this out to .NET 10.

antonfirsov committed:
move log message implementation into LogHelper, introduce an environment variable for the fallback switch

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/RedactedLogValueIntegrationTest.cs:
```diff
@@ -16,6 +16,15 @@ namespace Microsoft.Extensions.Http.Logging
 {
     public class RedactedLogValueIntegrationTest
     {
+        private static class EventIds
```
I think we should be able to access LogHelper from the tests -- we have InternalsVisibleTo set up here https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Http/src/Properties/InternalsVisibleTo.cs#L6

antonfirsov replied:
I'm not big fan of reusing product logic in it's own tests, it might hurt some aspects of the testing. Eg. if you accidentally change the id in the test will not detect the breaking change. Or is there another reason for the dependency I'm missing?

antonfirsov committed:
redact query string with '*'

antonfirsov commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,182 @@
+            public static readonly EventId PipelineStart = new EventId(100, "RequestPipelineStart");
+            public static readonly EventId PipelineEnd = new EventId(101, "RequestPipelineEnd");
+
+            public static readonly EventId RequestPipelineRequestHeader = new EventId(102, "RequestPipelineRequestHeader");
```
Is it intentional that `RequestHeader` and `RequestPipelineRequestHeader` only differ in their names?

antonfirsov replied:
>  they are used strictly for different loggers/categories

Does this mean, that it is unlikely that it will confuse log consumer (sink) logic in practice, despite `EventId.Equals()` and `==` only comparing the `Id` values?

antonfirsov commented on src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/LoggingUriOutputTests.cs:
```diff
@@ -16,156 +16,198 @@ namespace Microsoft.Extensions.Http.Tests.Logging
+            { "http://q.app/foo", "http://q.app/foo" },
+            { "http://q.app/foo?", "http://q.app/foo?" },
+            { "http://q.app/foo?XXX", "http://q.app/foo?*" },
+            { "http://q.app/a/b/c?a=b%20c&x=1", "http://q.app/a/b/c?*" },
```
Unlike with headers, where AFAIK we only replace the hader values, we replace the entire query string with a `*`.

We could log `?a=*&x=*` here, but it would come with a parsing price I assume we don't want to pay for the default log level.

MihaZupan commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,182 @@
+            }
+
+#if NET
+            return string.Create(queryOffset + 2, (uriString, queryOffset), (result, s) =>
+            {
+                ReadOnlySpan<char> withoutQuery = s.uriString.AsSpan(0, s.queryOffset + 1);
+                withoutQuery.CopyTo(result);
+                Debug.Assert(result[^2] == '?');
+                result[^1] = '*';
+            });
```
```suggestion
            return $"{uriString.AsSpan(0, queryOffset + 1)}*";
```

We could probably come up with some incantation of `$"{uri.Scheme}://{uri.Foo}{someSpan}*"` to save 1 allocation if we care enough.

antonfirsov committed:
suggestion

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>

antonfirsov committed:
rename switch

CarnaViire reviewed (dismissed):
Some nits (leftover namings mostly), otherwise LGTM, thanks!

antonfirsov committed:
suggestions

Co-authored-by: Natalia Kondratyeva <knatalia@microsoft.com>

antonfirsov committed:
Merge branch 'main' into querystring-01

antonfirsov committed:
Merge branch 'querystring-01' of https://github.com/antonfirsov/runtime into querystring-01

antonfirsov commented on src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Logging/LoggingUriOutputTests.cs:
```diff
@@ -16,156 +16,198 @@ namespace Microsoft.Extensions.Http.Tests.Logging
+                }
+            }
+
+            result.Add(false, false, "EnvVarTrue");
```
That would double the test-space, IMO unnecessarily. Testing the various global configuration modes on the async API only feels enough.

antonfirsov committed:
adjustments

antonfirsov committed:
handle fragment

antonfirsov committed:
remove userinfo from absolute Uri-s

antonfirsov commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,197 @@
+                return null;
+            }
+
+            string uriString = uri.IsAbsoluteUri
+                ? uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.UserInfo, UriFormat.UriEscaped)
+                : uri.ToString();
```
We also have to remove userinfo. In order to also remove it from relative Uri-s we would need to duplicate a non-trivial amount of Uri parsing code in the package. Given that the prence of relative Uri-s here is an edge-case  (one would need to use `LoggingHttpMessageHandler` manually with `HttpMessageInvoker`), I'm not sure if we should bother doing that.

@blowdart @MihaZupan thoughts?

Marking this as `NO-MERGE` until this is resolved.

antonfirsov replied:
Alternatively, we can omit logging relative Uri-s, IMHO that would be the simplest thing to deal with this edge-case.

antonfirsov replied:
I pushed a change that redacts relative URIs unconditionally, PTAL.

blowdart replied:
People still use username/password in URIs? It's long deprecated now and most modern servers don't even do anything with it :)

CarnaViire replied:
> It's long deprecated now

But isn't only `username:password` combination deprecated? 🤔🤔🤔 and `usename`-only still can be used and should be supported? 🤔 

(I judge only by what I've seen in the RFC though, so I might be wrong)

<details>

https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1

I understand the following _only_ applies to the "user:password" format:

```
   Use of the format "user:password" in the userinfo field is
   deprecated.  Applications should not render as clear text any data
   after the first colon (":") character found within a userinfo
   subcomponent unless the data after the colon is the empty string
   (indicating no password).  Applications may choose to ignore or
   reject such data when it is received as part of a reference and
   should reject the storage of such data in unencrypted form.  The
   passing of authentication information in clear text has proven to be
   a security risk in almost every case where it has been used.
```

And the following implies that `usename`-only is ok if the user wants to see it:

```
   Applications that render a URI for the sake of user feedback, such as
   in graphical hypertext browsing, should render userinfo in a way that
   is distinguished from the rest of a URI, when feasible.
```

</details>

blowdart replied:
So talking to Anton in teams I believe the sticking point here is for URIs with relative schemes, eg. "//user:pass@example/foo/bar".

But does the http client support those? They really don't make sense outside of a containing web page where the protocol is taken from the page that contains such links?

antonfirsov replied:
I don't think that's the case. According to the [4.2 Relative Reference](https://datatracker.ietf.org/doc/html/rfc3986#section-4.2).

> A relative reference takes advantage of the hierarchical syntax (Section 1.2.3) to express a URI reference relative to the name space of another hierarchical URI.

Which seems to be the definition of what we call relative Uri. I'm still having trouble to imagine how can relative Uri-s contain UserInfo in practice. @MihaZupan do you have any insight?

MihaZupan replied:
> do you have any insight?

Only that when you first asked me about it, I had to double-check that this is a thing, and that `System.Uri` supports it.
You can only supply the user info when also supplying the authority, and that is only possible with the `"//" authority path-abempty` syntax, which even the RFC calls out as rare
> A relative reference that begins with two slash characters is termed a network-path reference; such references are rarely used.

This is also a format where it may be ambiguous as to whether you mean a relative network-path reference vs. an absolute UNC path depending on whether you allow `Uri` to use `UriKind.Relative` or not.
```c#
Console.WriteLine(new Uri("//domain/path", UriKind.Absolute).IsAbsoluteUri);
Console.WriteLine(new Uri("//domain/path", UriKind.Relative).IsAbsoluteUri);
```

---

> But does the http client support those?

Our code that sends the request (SocketsHttpHandler) only cares about absolute Uris.

But technically a user can create a custom `HttpMessageHandler` chain that accepts a relative Uri and then replaces it with an absolute one sometime later before calling into `SocketsHttpHandler`.
And inside this chain they _could_ call into `LoggingHttpMessageHandler` before they've replaced the Uri with an absolute one.

---

TL;DR; IMO it seems like a pretty obscure thing for someone to do, in which case they should just... not do that :)

blowdart replied:
I guess someone could be writing a .NET based browser and wants protocol relative, rather than just path relative. What a mess. URIs with UNC formats shouldn't accept username/passwords though, which presumably would help disambiguate.

So, given the format which requires user:pass@ in the first segment after an optional scheme, can you not just pattern match on that and strip the segment?

antonfirsov replied:
I think the simplest thing is to stick with the strategy of redacting relative Uri-s entirely. If anyone will be bitten by this, e557330 now enables opting out from redaction entirely & they can open an issue for us to revisit this decision.

Scream if you disagree :)

CarnaViire commented:
(marking as draft to prevent accidental merge -- NO-MERGE isn't sufficient)

antonfirsov committed:
Redact relative Uri unconditionally

CarnaViire commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -151,9 +151,16 @@ public static void LogRequestPipelineEnd(this ILogger logger, HttpResponseMessag
-            string uriString = uri.IsAbsoluteUri
-                ? uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.UserInfo, UriFormat.UriEscaped)
-                : uri.ToString();
+            if (!uri.IsAbsoluteUri)
```
This should be after `s_disableUriQueryRedaction` check

CarnaViire replied:
Because, to my understanding, the switch should return everything to the old behavior

CarnaViire replied:
I don't care that much about not being able to do a precise redaction of a relative uri, what I'm not happy about is introducing a breaking change without a straightforward way to opt-out.

But whatever. While I was able to find some interesting usecases for UserInfo in Uri in general, but for a plain HttpClientFactory case it _should_ be rare. Matching the OTel logic sounds ok-ish... (but we don't fully match their format, do we?)

MihaZupan commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,204 @@
+                return "*";
+            }
+
+            string uriString = uri.GetComponents(UriComponents.AbsoluteUri & ~UriComponents.UserInfo, UriFormat.UriEscaped);
```
I just remembered this won't actually work since Uris created with `DangerousDisablePathAndQueryCanonicalization` won't let you ask for the path/query via `GetComponents` (we might have to rethink that).

A low-alloc version would look something like this:
```c#
private static string GetRedactedUriString(Uri? uri)
{
    if (uri is null) return null;
    
    if (!uri.IsAbsoluteUri) return "*";

    string pathAndQuery = uri.PathAndQuery;
    int queryIndex = s_disableUriQueryRedaction ? -1 : pathAndQuery.IndexOf('?');

    return (queryIndex >= 0, uri.IsDefaultPort) switch
    {
        (true, true) => $"{uri.Scheme}://{uri.Host}{pathAndQuery.AsSpan(0, queryIndex + 1)}*",
        (true, false) => $"{uri.Scheme}://{uri.Host}:{uri.Port}{pathAndQuery.AsSpan(0, queryIndex + 1)}*",
        (false, true) => $"{uri.Scheme}://{uri.Host}{pathAndQuery}",
        (false, false) => $"{uri.Scheme}://{uri.Host}:{uri.Port}{pathAndQuery}"
    };
}
```

Note that this also removes the fragment. Do we care about it? It's unlikely to be set and won't be sent over the wire anyway.

antonfirsov replied:
>Do we care about it? It's unlikely to be set and won't be sent over the wire anyway.

I would prefer to have consistency between the redaction here and redaction for `url.full` implemented in https://github.com/dotnet/runtime/pull/104251, where I wanted to be compatible with the [OTel logic](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/4c69afa2fbb3b9d84b45bd73dcfad43331516e69/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpTagHelper.cs#L21) which does emit the fragment. But given that our query redaction is already incompatible we can probably omit fragment as well. I will ping Vishwesh on this.

antonfirsov committed:
Merge branch 'main' into querystring-01

antonfirsov committed:
Harmonize redaction:

- Use a shared switch to opt-out from redaction entirely (System.Net.Http.DisableUriRedaction)
- Simplify and optimize redaction logic, always omit Fragment when redacting
- Adjust and consolidate tests

antonfirsov commented:
@CarnaViire @MihaZupan with e557330 this is ready for review again.

antonfirsov committed:
AggressiveInlining

CarnaViire reviewed (approved):
LGTM, thanks!

MihaZupan commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,192 @@
+
+        public static void LogRequestStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            // We check here to avoid allocating in the GetUriString call unnecessarily
```
```suggestion
            // We check here to avoid allocating in the GetRedactedUriString call unnecessarily
```

MihaZupan commented on src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs:
```diff
@@ -0,0 +1,192 @@
+            }
+        }
+
+        public static IDisposable? BeginRequestPipelineScope(this ILogger logger, HttpRequestMessage request)
+        {
+            return _beginRequestPipelineScope(logger, request.Method, GetRedactedUriString(request.RequestUri));
+        }
+
+        public static void LogRequestPipelineStart(this ILogger logger, HttpRequestMessage request, Func<string, bool> shouldRedactHeaderValue)
+        {
+            _requestPipelineStart(logger, request.Method, GetRedactedUriString(request.RequestUri), null);
```
We were using `AbsoluteUri` before, which would get cached by the `Uri`.
With `GetRedactedUriString` that's not the case, and we might be allocating it three times. Can we avoid that here?

antonfirsov replied:
I was able to save one of the `LoggingScopeHttpMessageHandler` allocations in [4b1d04](https://github.com/dotnet/runtime/pull/103769/commits/74b1d040d4a48cf08c201d4936f80d766aa1b9c0).

Unfortunately, there is no sane way that would allow us to flow the uri string further to `LoggingHttpMessageHandler`.

MihaZupan replied:
It's unfortunate, but makes sense. Thanks.

antonfirsov committed:
save 1 allocation in LoggingScopeHttpMessageHandler

antonfirsov committed:
Update src/libraries/Microsoft.Extensions.Http/src/Logging/LogHelper.cs

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>

antonfirsov commented:
The `LibraryImportGenerator.Tests` failure is unrelated.

liveans commented:
Breaking change issue is created for this: https://github.com/dotnet/docs/issues/42792
